Sources/Operators/ConcatMap.swift

```diff
@@ -0,0 +1,163 @@
+
+        public init(
+            upstream: Upstream,
+            transform: @escaping Transform
+        ) {
+            self.upstream = upstream
+            self.transform = transform
+        }
+
+        public func receive<S: Subscriber>(subscriber: S)
+        where Output == S.Input, Failure == S.Failure {
+            subscriber.receive(
+                subscription: Subscription(
+                    upstream: upstream,
+                    transform: transform,
+                    downstream: subscriber
+                )
+            )
+        }
+    }
+}
+
+// MARK: - Subscription
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+private extension Publishers.ConcatMap {
+    final class Subscription<Downstream: Subscriber>: Combine.Subscription where Downstream.Input == Output, Downstream.Failure == Failure {
+        private var sink: Sink<Downstream>?
+
+        init(
+            upstream: Upstream,
+            transform: @escaping Transform,
+            downstream: Downstream
+        ) {
+            self.sink = Sink(
+                upstream: upstream,
+                downstream: downstream,
+                transform: { transform($0) }
+            )
+        }
+
+        func request(_ demand: Subscribers.Demand) {
+            sink?.demand(demand)
+        }
+
+        func cancel() {
+            sink = nil
+        }
+    }
+}
+
+// MARK: - Sink
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+private extension Publishers.ConcatMap {
+    final class Sink<Downstream: Subscriber>: CombineExt.Sink<Upstream, Downstream> where Downstream.Input == Output, Downstream.Failure == Failure {
+        private let lock = NSRecursiveLock()
+        private let transform: Transform
+        private var activePublisher: NewPublisher?
+        private var bufferedPublishers: [NewPublisher]
+        private var cancellables: Set<AnyCancellable>
+
+        init(
+            upstream: Upstream,
+            downstream: Downstream,
+            transform: @escaping Transform
+        ) {
+            self.transform = transform
+            self.bufferedPublishers = []
+            self.cancellables = []
+            super.init(
+                upstream: upstream,
+                downstream: downstream,
+                transformFailure: { $0 }
+            )
+        }
+
+        override func receive(_ input: Upstream.Output) -> Subscribers.Demand {
+            let mapped = transform(input)
+            lock.lock()
+            defer { lock.unlock() }
+
+            if activePublisher == nil {
+                setActivePublisher(mapped)
+            } else {
+                bufferedPublishers.append(mapped)
+            }
+
+            return .unlimited
+        }
+
+        private func setActivePublisher(_ publisher: NewPublisher) {
+            lock.lock()
+            defer { lock.unlock() }
+            activePublisher = publisher
+
+            publisher.sink(
```

Comment:
I'm not sure I'm a fan of using sink inside a custom publisher. I mean, you're alraedy holding your own subscription etc so why use a built-in subscriber that creates a subscription to do this ?

Reply:
Problem is that we have an outer publisher of publishers and we want to flatten the value of the inner publisher into one publisher of values.

I checked how [RxSwift solves this](https://github.com/ReactiveX/RxSwift/blob/main/RxSwift/Observables/Merge.swift#L133) and it relies on `MergeLimitedSink` which subscribes a `MergeLimitedSinkIter` (inner sink) to each inner publisher and then checks if the outer sink [has a 'next' publisher](https://github.com/ReactiveX/RxSwift/blob/main/RxSwift/Observables/Merge.swift#L172) when the inner publisher completes.

The only benefit that I see from this is that it reuses 'active sinks' and subscribes them directly to the next publisher. Maybe you see some other benefits in this approach? I'm not 100% familiar with the RxSwift codebase.

Reply:
Just discovered that having an outer / inner sink would solve the whole uncertainty around which demand to return. I'll rewrite it to look a bit more like the RxSwift equivalent.